crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -223,42 +237,171 @@ impl<'flox> Publish<'flox, NixAnalysis> {
     }
 
     /// Write snapshot to catalog and push to origin
-    pub async fn push_catalog(self) -> Result<(), PublishError> {
-        let url = self.publish_flake_ref.clone_url();
-        let repo_dir = tempfile::tempdir_in(&self.flox.temp_dir)
-            .unwrap()
-            .into_path(); // todo catch error
-        let catalog = <Git as GitProvider>::clone(url, &repo_dir, false)
-            .await
-            .unwrap(); // todo: catch error
+    pub async fn push_snapshot(&self) -> Result<(), PublishError> {
+        let mut upstream_repo =
+            UpstreamRepo::clone_repo(self.publish_flake_ref.clone_url(), &self.flox.temp_dir)
+                .await?;
+        self.push_snapshot_to(&mut upstream_repo).await
+    }
+
+    /// Write snapshot to a catalog and push to 'origin'
+    ///
+    /// Internal method to test
+    async fn push_snapshot_to(&self, upstream_repo: &mut UpstreamRepo) -> Result<(), PublishError> {
+        let catalog = upstream_repo
+            .get_or_create_catalog(&self.flox.system)
+            .await?;
+        if let Ok(Some(_)) = catalog.get_snapshot(self.analysis()) {
+            Err(PublishError::SnapshotExists)?;
+        }
+        catalog.add_snapshot(self.analysis()).await?;
+        catalog.push_catalog().await?;
+        Ok(())
+    }
+}
 
-        if catalog.list_branches().await.unwrap() // todo: catch error
-            .into_iter().any(|info| info.name == "catalog")
+/// Representation of an exclusive clone of an upstream repo
+///
+/// [UpstreamRepo] and [UpstreamCatalog] ensure safe access to individual catalog branches.
+/// Every [UpstreamRepo] instance represents an exclusive clone
+/// and can only ever create a single [UpstreamCatalog] instance at a time.
+struct UpstreamRepo(Git);
+
+impl UpstreamRepo {
+    /// Clone an upstream repo
+    async fn clone_repo(
+        url: impl AsRef<str>,
+        temp_dir: impl AsRef<Path>,
+    ) -> Result<Self, PublishError> {
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(url.as_ref(), &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
```

Comment:
I was wondering today, will we support catalogs for different branches? E.g. if I have two branches nixpkgs-stable and nixpkgs-unstable in the same repo, should we create `nixpkgs-stable/catalog/{system}` and `nixpkgs-unstable/catalog/{system}` so there isn't a collision? No need to change that now, just thought I'd mention the question